File: backend/app/schemas/resume.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# --- Sub-schemas ---

class BasicsSchema(BaseModel):
    name: str = ""
    label: str = ""
    email: str = ""
    phone: str = ""
    url: str = ""
    summary: str = ""
    location: dict[str, str] = Field(default_factory=dict)
    profiles: list[dict[str, str]] = Field(default_factory=list)
    avatar: str = ""


class EducationItem(BaseModel):
    institution: str = ""
    area: str = ""
    study_type: str = ""
    start_date: str = ""
    end_date: str = ""
    score: str = ""
    courses: list[str] = Field(default_factory=list)


class WorkItem(BaseModel):
    company: str = ""
    position: str = ""
    website: str = ""
    start_date: str = ""
    end_date: str = ""
    summary: str = ""
    highlights: list[str] = Field(default_factory=list)


class SkillItem(BaseModel):
    name: str = ""
    level: str = ""
    keywords: list[str] = Field(default_factory=list)


class ProjectItem(BaseModel):
    name: str = ""
    description: str = ""
    url: str = ""
    highlights: list[str] = Field(default_factory=list)
    keywords: list[str] = Field(default_factory=list)
    start_date: str = ""
    end_date: str = ""


class CustomSection(BaseModel):
    id: str = ""
    title: str = ""
    items: list[dict[str, Any]] = Field(default_factory=list)


class StyleConfig(BaseModel):
    font_family: str = "'Noto Sans SC', sans-serif"
    font_size: float = 14
    line_height: float = 1.6
    margin_top: float = 20
    margin_bottom: float = 20
    margin_left: float = 24
    margin_right: float = 24
    section_gap: float = 16
    primary_color: str = "#00f0ff"
    accent_color: str = "#ff2d6d"
    text_color: str = "#e0e0e0"
    background_color: str = "#0a0a0f"
    heading_size: float = 18
    page_size: str = "A4"


# --- API schemas ---

class ResumeCreate(BaseModel):
    title: str
    slug: str
    template_id: str = "classic"
    basics: BasicsSchema | None = None
    education: list[EducationItem] | None = None
    work: list[WorkItem] | None = None
    skills: list[SkillItem] | None = None
    projects: list[ProjectItem] | None = None
    awards: list[dict[str, Any]] | None = None
    languages: list[dict[str, str]] | None = None
    interests: list[dict[str, Any]] | None = None
    custom_sections: list[CustomSection] | None = None
    style_config: StyleConfig | None = None


class ResumeUpdate(BaseModel):
    title: str | None = None
    template_id: str | None = None
    basics: BasicsSchema | None = None
    education: list[EducationItem] | None = None
    work: list[WorkItem] | None = None
    skills: list[SkillItem] | None = None
    projects: list[ProjectItem] | None = None
    awards: list[dict[str, Any]] | None = None
    languages: list[dict[str, str]] | None = None
    interests: list[dict[str, Any]] | None = None
    custom_sections: list[CustomSection] | None = None
    style_config: StyleConfig | None = None


class ResumeResponse(BaseModel):
    id: str
    title: str
    slug: str
    template_id: str
    basics: dict[str, Any] | None
    education: list[dict[str, Any]] | None
    work: list[dict[str, Any]] | None
    skills: list[dict[str, Any]] | None
    projects: list[dict[str, Any]] | None
    awards: list[dict[str, Any]] | None
    languages: list[dict[str, str]] | None
    interests: list[dict[str, Any]] | None
    custom_sections: list[dict[str, Any]] | None
    style_config: dict[str, Any] | None
    source: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ResumeListItem(BaseModel):
    id: str
    title: str
    slug: str
    template_id: str
    source: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/resume.ts
export interface Basics {
  name: string;
  label: string;
  email: string;
  phone: string;
  url: string;
  summary: string;
  location: Record<string, string>;
  profiles: SocialProfile[];
  avatar: string;
}

export interface SocialProfile {
  network: string;
  username: string;
  url: string;
}

export interface EducationItem {
  institution: string;
  area: string;
  study_type: string;
  start_date: string;
  end_date: string;
  score: string;
  courses: string[];
  /** Markdown 详情描述 */
  description: string;
}

export interface WorkItem {
  company: string;
  /** 部门名称 */
  department: string;
  position: string;
  website: string;
  start_date: string;
  end_date: string;
  /** Markdown 工作详情（用户自由编写） */
  description: string;
  /** 公司 Logo（Base64 data URL） */
  logo: string;
}

export interface SkillItem {
  /** Markdown 技能描述（用户自由编写） */
  content: string;
}

export interface ProjectItem {
  name: string;
  description: string;
  url: string;
  highlights: string[];
  keywords: string[];
  start_date: string;
  end_date: string;
}

export interface AwardItem {
  title: string;
  date: string;
  awarder: string;
  summary: string;
}

export interface LanguageItem {
  language: string;
  fluency: string;
}

export interface InterestItem {
  name: string;
  keywords: string[];
}

export interface CustomSectionItem {
  title: string;
  subtitle: string;
  date: string;
  description: string;
  highlights: string[];
}

export interface CustomSection {
  id: string;
  title: string;
  items: CustomSectionItem[];
}

/** Structured heading decoration config */
export interface HeadingStyle {
  /** 标题字体 (留空 = 跟随正文) */
  font_family: string;
  /** 标题字号 px */
  font_size: number;
  /** 标题颜色 (留空 = 使用主色) */
  color: string;
  /** 是否显示左侧竖线 */
  left_bar: boolean;
  /** 左侧竖线颜色 (留空 = 使用主色) */
  left_bar_color: string;
  /** 下划线样式: 'none' | 'solid' | 'dashed' | 'dotted' | 'double' */
  underline: 'none' | 'solid' | 'dashed' | 'dotted' | 'double';
  /** 下划线颜色 (留空 = 使用主色) */
  underline_color: string;
  /** 标题背景色 (留空 = 无背景) */
  background: string;
}

export const DEFAULT_HEADING_STYLE: HeadingStyle = {
  font_family: '',
  font_size: 18,
  color: '',
  left_bar: false,
  left_bar_color: '',
  underline: 'solid',
  underline_color: '',
  background: '',
};

export interface StyleConfig {
  font_family: string;
  font_size: number;
  line_height: number;
  margin_top: number;
  margin_bottom: number;
  margin_left: number;
  margin_right: number;
  section_gap: number;
  primary_color: string;
  text_color: string;
  background_color: string;
  heading_size: number;
  page_size: string;
  /** @deprecated 使用 heading_style 代替 */
  custom_css: string;
  /** 结构化标题样式 */
  heading_style: HeadingStyle;
  /** Logo 是否显示边框 */
  logo_border: boolean;
  /** Logo 边框圆角 px */
  logo_border_radius: number;
}

/** A snapshot of a resume at a point in time */
export interface ResumeVersion {
  id: string;
  timestamp: number;
  label: string;
  data: ResumeData;
}

/** Controls which sections are visible and their order */
export interface SectionVisibility {
  summary: boolean;
  work: boolean;
  education: boolean;
  skills: boolean;
  projects: boolean;
  awards: boolean;
  languages: boolean;
  interests: boolean;
}

export const DEFAULT_SECTION_VISIBILITY: SectionVisibility = {
  summary: true,
  work: true,
  education: true,
  skills: true,
  projects: true,
  awards: false,
  languages: false,
  interests: false,
};

/** Section ordering: the display order of sections */
export type SectionKey = keyof SectionVisibility;

export const DEFAULT_SECTION_ORDER: SectionKey[] = [
  'summary',
  'work',
  'projects',
  'education',
  'skills',
  'awards',
  'languages',
  'interests',
];

export interface ResumeData {
  id: string;
  title: string;
  slug: string;
  template_id: string;
  basics: Basics;
  education: EducationItem[];
  work: WorkItem[];
  skills_text: string;
  projects: ProjectItem[];
  awards: AwardItem[];
  languages: LanguageItem[];
  interests: InterestItem[];
  custom_sections: CustomSection[];
  style_config: StyleConfig;
  section_visibility: SectionVisibility;
  section_order: SectionKey[];
  source?: string;
  created_at: string;
  updated_at: string;
  sort_order: number;
  versions: ResumeVersion[];
}

export type TemplateId = 'classic' | 'modern' | 'minimal' | 'professional' | 'creative' | 'cyberpunk';

export const DEFAULT_STYLE_CONFIG: StyleConfig = {
  font_family: "'Noto Sans SC', 'Inter', sans-serif",
  font_size: 14,
  line_height: 1.6,
  margin_top: 20,
  margin_bottom: 20,
  margin_left: 24,
  margin_right: 24,
  section_gap: 16,
  primary_color: '#00f0ff',
  text_color: '#1a1a1a',
  background_color: '#ffffff',
  heading_size: 18,
  page_size: 'A4',
  custom_css: '',
  heading_style: { ...DEFAULT_HEADING_STYLE },
  logo_border: true,
  logo_border_radius: 6,
};

export const DEFAULT_BASICS: Basics = {
  name: '',
  label: '',
  email: '',
  phone: '',
  url: '',
  summary: '',
  location: {},
  profiles: [],
  avatar: '',
};

export function createEmptyResume(): ResumeData {
  const now = new Date().toISOString();
  return {
    id: crypto.randomUUID(),
    title: '未命名简历',
    slug: `resume-${Date.now()}`,
    template_id: 'classic',
    basics: { ...DEFAULT_BASICS },
    education: [],
    work: [],
    skills_text: '',
    projects: [],
    awards: [],
    languages: [],
    interests: [],
    custom_sections: [],
    style_config: { ...DEFAULT_STYLE_CONFIG },
    section_visibility: { ...DEFAULT_SECTION_VISIBILITY },
    section_order: [...DEFAULT_SECTION_ORDER],
    created_at: now,
    updated_at: now,
    sort_order: 0,
    versions: [],
  };
}
